fendermint/vm/interpreter/src/fvm/exec.rs

```diff
@@ -212,21 +219,47 @@ where
             PowerUpdates::default()
         };
 
-        // check for upgrades in the upgrade_scheduler
-        let chain_id = state.chain_id();
-        let block_height: u64 = state.block_height().try_into().unwrap();
-        if let Some(upgrade) = self.upgrade_scheduler.get(chain_id, block_height) {
-            // TODO: consider using an explicit tracing enum for upgrades
-            tracing::info!(?chain_id, height = block_height, "Executing an upgrade");
-
-            // there is an upgrade scheduled for this height, lets run the migration
-            let res = upgrade.execute(&mut state).context("upgrade failed")?;
-            if let Some(new_app_version) = res {
-                state.update_app_version(|app_version| {
-                    *app_version = new_app_version;
-                });
-
-                tracing::info!(app_version = state.app_version(), "upgraded app version");
+        if let Some(upgrade_info) = self.upgrade_schedule.get(state.block_height()) {
+            tracing::info!(
+                app_version = state.app_version(),
+                new_app_version = upgrade_info.new_app_version,
+                height = state.block_height(),
+                upgrade_info = ?upgrade_info,
+                "upgrade scheduled"
+            );
+
+            match self.upgrades.get(upgrade_info.new_app_version) {
+                Some(upgrade) => {
+                    let new_app_version = upgrade.execute(&mut state).context("upgrade failed")?;
+
+                    state.update_app_version(|app_version| {
+                        *app_version = new_app_version;
+                    });
+
+                    tracing::info!(app_version = state.app_version(), "upgrade successful");
+                }
+                None => {
+                    tracing::warn!(
+                        height = state.block_height(),
+                        upgrade_info = ?upgrade_info,
+                        "upgrade not found"
+                    );
+
+                    if upgrade_info.required {
+                        // mark the node as frozen
+                        IS_FROZEN.store(true, Ordering::Relaxed);
+
+                        // sleep forever, we can't proceed any further using our current fendermint version
```

Comment:
Maybe this can be in `begin` instead of `end` to minimise work; like if we know we won't be able to finish the block, why even start executing it?